src/main/kotlin/org/rust/lang/refactoring/generateConstructor/GenerateConstructorHandler.kt

```diff
@@ -0,0 +1,119 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.lang.refactoring.generateConstructor
+
+
+import com.intellij.codeInsight.CodeInsightActionHandler
+import com.intellij.codeInsight.actions.CodeInsightAction
+import com.intellij.lang.LanguageCodeInsightActionHandler
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.application.runWriteAction
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiFile
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.ancestorOrSelf
+import org.rust.lang.core.psi.ext.isTupleStruct
+import org.rust.openapiext.checkWriteAccessAllowed
+
+class GenerateConstructorAction : CodeInsightAction() {
+    override fun getHandler(): CodeInsightActionHandler = generateConstructorHandler
+    override fun isValidForFile(project: Project, editor: Editor, file: PsiFile): Boolean =
+        GenerateConstructorHandler().isValidFor(editor, file)
+
+    private val generateConstructorHandler = GenerateConstructorHandler()
+}
+
+class GenerateConstructorHandler : LanguageCodeInsightActionHandler {
+    override fun isValidFor(editor: Editor, file: PsiFile): Boolean =
+        getStructItem(editor, file) != null
+
+    override fun startInWriteAction() = false
+    private fun getStructItem(editor: Editor, file: PsiFile): RsStructItem? =
+        file.findElementAt(editor.caretModel.offset)?.ancestorOrSelf()
+
+    override fun invoke(project: Project, editor: Editor, file: PsiFile) {
+        val structItem = getStructItem(editor, file) ?: return
+        generateConstructorBody(structItem, editor)
+    }
+
+
+    private fun generateConstructorBody(structItem: RsStructItem, editor: Editor) {
+        check(!ApplicationManager.getApplication().isWriteAccessAllowed)
+        val chosenFields = showConstructorArgumentsChooser(structItem, structItem.project) ?: return
+        runWriteAction {
+            insertNewConstructor(structItem, chosenFields, editor)
+        }
+    }
+
+    private fun insertNewConstructor(structItem: RsStructItem, selectedFields: List<ConstructorArguments>, editor: Editor) {
+        checkWriteAccessAllowed()
+        val psiFactory = RsPsiFactory(editor.project ?: return)
+        var expr = psiFactory.createInherentImplItem(structItem.name ?: return, structItem.typeParameterList
```

Comment:
Please, try to avoid return inside long expressions. Consider extracting variables before passing to a function.